11-lorawan/test_spec11.py

```diff
@@ -0,0 +1,118 @@
+import time
+import pytest
+from testutils.shell import GNRCLoRaWANSend, ifconfig, lorawan_netif
+from riotctrl_shell.netif import Ifconfig
+import base64
+import json
+
+APP = 'examples/gnrc_lorawan'
+pytestmark = pytest.mark.rc_only()
+
+APP_PAYLOAD = "This is RIOT!"
+DOWNLINK_PAYLOAD = "VGhpcyBpcyBSSU9U"
+APP_PORT = 2
+RX2_DR = 3
+
+
+class Shell(Ifconfig, GNRCLoRaWANSend):
+    pass
+
+
+def run_lw_test(node, client, userdata, iface, app_id, dev_id):
+    # Disable confirmable messages
+    node.ifconfig_flag(iface, "ack_req", enable=False)
+
+    # Push a downlink message to the TTN server
+    dl_data = {"payload_raw": DOWNLINK_PAYLOAD, "port": APP_PORT,
+               "confirmed": True}
+
+    client.publish("{}/devices/{}/down".format(app_id, dev_id),
+                   json.dumps(dl_data))
+
+    # Send a message. The send function will return True if the downlink is
+    # receives (as expected)
+    assert node.send(iface, APP_PAYLOAD) is True
+    time.sleep(3)
```

Comment:
This could be better synchronized now by having a condition or something?

in `TTNClient.__init__()`:

```py
        client = mqtt.Client()
        self.client = client
        self.userdata = {}
        self._msg_received = threading.Condition()
```

in `TTNClient.__enter__()`:

```py
    def __enter__(self):
        def _on_message(client, userdata, msg):
            with self._msg_received:
                on_message(client, userdata, msg)
                self._msg_received.notify_all()
        
        client.on_connect = on_connect
        client.on_message = _on_message
        self.client.user_data_set(self.userdata)
        self.client.tls_set()
        password = get_required_envvar("LORAWAN_DL_KEY")
        self.client.username_pw_set(APP_ID, password=password)
        self.client.connect('eu.thethings.network', 8883, 60)
        self.client.loop_start()
        return self
```

and in a new method

```py
def wait_for_data(self):
    self._msg_received.wait()
```

Reply:
(`client.wait_for_data()` could then be called here instead of `time.sleep(3)` if that wasn't clear)

Reply:
Ping?

Reply:
on it :+1:

Reply:
note this `time.sleep` is not a synchronization timer but rather a duty cycle enforcer. I renamed the variable to make it more clear